File: Hash Map/Subarray of sum equals K/optimal.py
class Solution:
	def subarraySum(self, nums: List[int], k: int) -> int:


		res = 0
		currSum = 0
		preSums = { 0 : 1 }

		for x in nums:

			currSum += x

			diff = currSum - k

			res += preSums.get(diff, 0)

			preSums[currSum] = 1 + preSums.get(currSum, 0)

		return res

File: Hash Map/Subarray of sum equals K/optimal.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        
        
        int res = 0;
        
        int currsum = 0;
        
        int diff;
        
        HashMap<Integer,Integer> presum = new HashMap<>();
        
        
        presum.put(0,1);
        
        for(int i=0;i<nums.length;i++){
            
            
            currsum += nums[i];
            
            diff = currsum - k;
            
            if(presum.containsKey(diff)){
                res += presum.get(diff);
            }
            if(presum.containsKey(currsum)){
                presum.put(currsum, 1 + presum.get(currsum));
            }else{
                presum.put(currsum, 1);
            }
            
        }
        
        return res;
        
    }
}